### DISCUSSION TITLE
How to set context-registry-service:read-only for a spa application?

### DISCUSSION THREAD

**Discussion by PeterJBassett** [iTwin/itwinjs-core#2130] [Q&A] [answered]
Hello, I have a bridge application test harness which queries the api using GET /contextregistry/v2.5/Repositories/BentleyCONNECT--Main/ConnectedContext/Project?$select=*&$filter=Name+eq+%27Bentley-FutureOn%20Connector%20Handover%27

This is failing with the error "Authorization scheme is not supported".

I believe that this is caused by a change I made to the scopes defined for my spa application.
I believe that I need to add the context-registry-service:read-only to the scopes. However the Apps portal (https://developer.bentley.com/my-apps) no longer seems to be able to add this scope.

**Comment by calebmshafer** [comment, marked as answer]
@PeterJBassett we recently made a change to the developer portal to try and simplify the set of scopes and managing access tokens, as we had two different ways of accessing APIs which lead to some confusion.

The "Authorization scheme is not supported" is due to a mis-match in the OIDC authority that is issuing the token. I'll avoid getting into the details just yet, see below if you're interested.

To fix the issue, I'd recommend ensuring you have the `itwinjs` scope added to your client and add it to the list of scopes requested by the app (also remove the `context-registry-service:read-only` scope). Then you will also want to set this env var for your app, `imjs_itwin_platform_authority=https://ims.bentley.com`.  With those changes, you should now be able to make the request to `contextregistry` url.

tldr;

The iTwin Platform has two supported OIDC Authorities for requesting an access token; `https://imsoidc.bentley.com` and `https://ims.bentley.com`.

The `imsoidc` based url has been the main one used over the past few years for iTwin.js and is how you would get a token that allowed contacting all of the services required to make your application (I'll refer to these as the initial APIs). This authority allowed you to get an access token with the following set of scopes;

- `rbac-user:external-client`
- `projectwise-share`
- `imodeljs-router`
- `general-purpose-imodeljs-backend`
- `product-setting-service`
- `context-registry-service:read-only`
- `imodelhub`
- `urlps-third-party`
- `reality-data:read`

With the new [iTwin Platform APIs](https://developer.bentley.com/apis/), released relatively recently, a new OIDC authority has been introduced as well, `ims.bentley.com`. The new authority, initially, provided you an access token that can be used with all of the new iTwin Platform APIs but did not allow you to contact any of "initial APIs" that iTwin.js still uses.

This lead to a requirement to know which API you wanted to call and which authority needed to be used to request an access token for that API.

Due to this complication being rather cumbersome for a developer to have to manage, we introduced a new way of getting an access token for the "initial APIs". We added the `itwinjs` scope as a supported scope on the new `ims.bentley.com` authority. This scope technically wraps up all of the scopes listed above and if you request an access token with just that single scope it will have access to call all of the "initial APIs".

So now, the recommended workflow is to always use the new authority (`ims.bentley.com`) and the `itwinjs` scope to call of the "initial APIs". This will allow you to remove from your list any of the 9 above you were requesting and then once you start using the new APIs, it can all be handled with a single access token rather than two.

**Reply by PeterJBassett** [reply]
Hello Caleb, thank you for your reply. Unfortunately, I have still not been able to fix the problem.
I have now configured my SPA application to enable all of the APIs that are possible to add, so the scope of my app is now `"email changedelements:modify issues:modify issues:read changedelements:read openid profile organization projects:read imodels:read storage:read storage:modify itwinjs imodels:modify projects:modify"`

However when I try to run a bridge synchronisation I get a 401 unauthorized error:
```
backend ResponseError: Unauthorized
backend     at Function.parse (/app/node_modules/@bentley/itwin-client/lib/Request.js:49:23)
backend     at parse (/app/node_modules/@bentley/itwin-client/lib/WsgClient.js:34:55)
backend     at parse (/app/node_modules/@bentley/imodelhub-client/lib/imodelhub/Errors.js:79:32)
backend     at Object.request (/app/node_modules/@bentley/itwin-client/lib/Request.js:289:29)
backend     at processTicksAndRejections (node:internal/process/task_queues:96:5)
backend     at async IModelBaseHandler.postInstance (/app/node_modules/@bentley/itwin-client/lib/WsgClient.js:275:21)
backend     at async BriefcaseHandler.create (/app/node_modules/@bentley/imodelhub-client/lib/imodelhub/Briefcases.js:194:21)
backend     at async Function.acquireNewBriefcaseId (/app/node_modules/@bentley/imodeljs-backend/lib/IModelHubBackend.js:128:27)
backend     at async Function.downloadBriefcase (/app/node_modules/@bentley/imodeljs-backend/lib/BriefcaseManager.js:166:89)
backend     at async BriefcaseDbBuilder.acquire (/app/node_modules/@bentley/imodel-bridge/lib/BridgeRunner.js:422:21) {
backend   errorNumber: 0,
backend   _getMetaData: undefined,
backend   description: 'cannot POST /imodelhub/sv1.1/Repositories/iModel--e8f65267-7494-4d03-b05f-1fd6085c3315/iModelScope/Briefcase/ (401)',
backend   _data: '',
backend   status: 401
backend }
```
Can you suggest a solution?

**Reply by aruniverse** [reply]
Just to confirm, you've updated the authority issuer to use the new ims.bentley.com authority right? 
If i'm not mistaken, Synchronization has its own scopes, `synchronization:modify` `synchronization:read`, that you don't have. In your request that fails, there should be a header `x-correlation-id`, if you could share that we can try to investigate on our end a bit.

**Reply by PeterJBassett** [reply]
Thanks for your reply - I have now added synchronization:modify synchronization:read, and still not working.
My application logs in twice to get two JWT tokens, one for synchronisations, and one for other APIs:
```
    // The Bentley authorisation service for synchronisation
    this.authURL = 'https://imsoidc.bentley.com/connect/authorize'

    // The Bentley authorisation service for other API access
    this.authApiURL = 'https://ims.bentley.com/connect/authorize'
```
Has this now changed?

**Reply by aruniverse** [reply]
Yes, you only need 1 token now. You only need to use the `ims` one, and can remove the use of the `imsoidc` one, Take a look at this announcement: https://github.com/imodeljs/imodeljs/discussions/2158.

**Reply by PeterJBassett** [reply]
Ok thank you that has fixed the synchronisation problem. 👍 
So the double login code that took time to be developed in my application will now have to be removed again. 👎

**Reply by aruniverse** [reply]
Sorry for the inconvenience, but this change should be beneficial long term, and simplify things.
